In [74]:
from sklearn.datasets import fetch_california_housing
import pandas as pd 

# Load the housing dataset
housing = fetch_california_housing()


X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

# Display the first few rows of the features and target
print(X.head())
print("\nTarget variable preview:")
print(y.head())


   MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25  

Target variable preview:
0    4.526
1    3.585
2    3.521
3    3.413
4    3.422
Name: med_house_value, dtype: float64


In [38]:
# Print the feature names and check for missing values
missing_datax= X.isnull().sum()
missing_datay= y.isnull().sum()
missing_datax_2= X.isnull().any()
#print(missing_datax)
print("missing data y : ",missing_datay)
print(missing_datax_2)

missing data y :  0
MedInc        False
HouseAge      False
AveRooms      False
AveBedrms     False
Population    False
AveOccup      False
Latitude      False
Longitude     False
dtype: bool


In [39]:
#Generate summary statistics (mean, min, max, etc.)
summary_stats = X.describe()
print("Summary Statistics for features:")

print(summary_stats)

Summary Statistics for features:
             MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude  
count  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704  
std       10.386050      2.135952      2.003532  
min        0.692308     32.540000   -124.350000  
25%  

In [40]:
#SPLit data into 80% training and 20% testing sets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Split the raw data (80% training, 20% testing)
X_train_raw, X_test_raw, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the linear regression model on unscaled data
lin_reg_raw = LinearRegression()
lin_reg_raw.fit(X_train_raw, y_train)

# Make predictions on the test set
y_pred_raw = lin_reg_raw.predict(X_test_raw)
print(y_pred_raw)



[0.71912284 1.76401657 2.70965883 ... 4.46877017 1.18751119 2.00940251]


In [41]:
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score

# Evaluate model performance
mse_raw = mean_squared_error(y_test, y_pred_raw)
rmse_raw = root_mean_squared_error(y_test, y_pred_raw)

r2_raw = r2_score(y_test, y_pred_raw)

print("Unscaled Data Model:")
print(f"Mean Squared Error: {mse_raw:.2f}")
print(f"Root Squared Error: {rmse_raw:.2f}")
print(f"R² Score: {r2_raw:.2f}")



Unscaled Data Model:
Mean Squared Error: 0.56
Root Squared Error: 0.75
R² Score: 0.58


In [42]:
# View our model's coefficients
print("Model Coefficients (Unscaled):")
print(pd.Series(lin_reg_raw.coef_,
                index=X.columns))
print("\nModel Intercept (Unscaled):")
print(pd.Series(lin_reg_raw.intercept_))

Model Coefficients (Unscaled):
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64

Model Intercept (Unscaled):
0   -37.023278
dtype: float64


In [43]:
coef_series = pd.Series(lin_reg_raw.coef_, index=X.columns)
intercept = lin_reg_raw.intercept_

print("Coefficients (Unscaled):")
print(coef_series)
print("\nIntercept:", intercept)

Coefficients (Unscaled):
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64

Intercept: -37.0232777060643


In [44]:
# This row's feature values will be multiplied by our coefficients.
row_index = 1  # for demonstration
row_features = X.iloc[row_index]  # features only
print("Feature values (Row", row_index, "):\n", row_features)

Feature values (Row 1 ):
 MedInc           8.301400
HouseAge        21.000000
AveRooms         6.238137
AveBedrms        0.971880
Population    2401.000000
AveOccup         2.109842
Latitude        37.860000
Longitude     -122.220000
Name: 1, dtype: float64


In [45]:
manual_prediction = (row_features * coef_series).sum() + intercept
print("\nManual Prediction for Row", row_index, ":", manual_prediction)



Manual Prediction for Row 1 : 3.9995200006664575


In [46]:
model_prediction = lin_reg_raw.predict([row_features])
print("Model Prediction from lin_reg_raw.predict():", model_prediction[0])

Model Prediction from lin_reg_raw.predict(): 3.9995200006664646


/opt/miniconda3/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [75]:
import pandas as pd

comparison = pd.DataFrame({
    "Actual": y_test.values,
    "Predicted": y_pred_raw
})

print(comparison.head(10))   # shows first 10 comparisons

    Actual  Predicted
0  0.47700   0.719123
1  0.45800   1.764017
2  5.00001   2.709659
3  2.18600   2.838926
4  2.78000   2.604657
5  1.58700   2.011754
6  1.98200   2.645500
7  1.57500   2.168755
8  3.40000   2.740746
9  4.46600   3.915615


###What does the R² score tell us about model performance?
- An R² of 0.58 means 58% of the variance in the outcome variable is explained by the model. it means the 58% of the errors that can be found in the use of the model can be explained by the same model. Which  means that the model is not being able to catch as much variance as would be ideal (a value closer to 1).

### Which features seem to have the strongest impact on predictions based on the model’s coefficients?
**Coeffecients in unscaled models tell us the direction of the effect that they may have in the model, isntead exactly the strength or true impact that they actually have.** 
> But in this case the strongest positive features are the following: 
-  Average Bedrooms, as houses have higher amount of bedrooms the more expensive they will be.
- Secondly one the anothe ehigh positive effect on the price of the house is the MedInc coeffecient as the areas that have a higher median income will be pricier than other locations. 

> For the coeffcients that have the highest negative effect ar ethe following: 
- As for longitud this immplies that as Going east toward inland California significantly reduces home prices.
- Going north (toward Northern California) corresponds to lower prices.
-(this could lead the the conclusion that certain locations will lead to lower hosue prices)

### How well do the predicted values match the actual values?
-  The model does an okay job predicting house prices. It gets the general pattern right, but the predictions don’t line up very closely with the actual values. This is reflected in an R² of 0.58 and an RMSE of 0.75, meaning the model is off by about 0.75 on average.



## Feature Selection and Simplified Model (25 points)

In [47]:
# Select three features from the dataset to build a simplified model. Explain your choice. (10 pts)
selected_features = ['MedInc', 'AveRooms', 'HouseAge']
X_selected = X[selected_features]
print("Selected Features:", selected_features)
# I chose 'MedInc' (Median Income) because it is likely a strong predictor of house values, as higher income areas tend 
# to have more expensive homes. 'AveRooms' (Average Rooms per House) is relevant because more rooms generally indicate a more valuable home. 
# 'HouseAge' is relevant because older homes may be less expensive than newer ones, depending on the condition and location.


Selected Features: ['MedInc', 'AveRooms', 'HouseAge']


In [67]:
# Train a new linear regression model using only these three features. (5 pts)
X_train_selected, X_test_selected, y_train, y_test = train_test_split(X_selected, y, test_size=0.2, random_state=42)
lin_reg_selected = LinearRegression()
lin_reg_selected.fit(X_train_selected, y_train)

# Make predictions on the test set
y_pred_selected = lin_reg_selected.predict(X_test_selected)
print(y_pred_selected)

[1.06791912 1.50634095 2.32862562 ... 4.33948872 1.71316594 1.77105121]


In [72]:
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score

mse_sel = mean_squared_error(y_test, y_pred_selected)
rmse_sel = root_mean_squared_error(y_test, y_pred_selected)
r2_sel = r2_score(y_test, y_pred_selected)

print("Simplified Model Metrics:")
print("MSE:", mse_sel)
print("RMSE:", rmse_sel)
print("R²:", r2_sel)

Simplified Model Metrics:
MSE: 0.6589108649336337
RMSE: 0.8117332473994359
R²: 0.49717158850807075


In [66]:
# View our model's coefficients
print("Model Coefficients (Unscaled):")
print(pd.Series(lin_reg_selected.coef_,
                index=X_selected.columns))
print("\nModel Intercept (Unscaled):")
print(pd.Series(lin_reg_selected.intercept_))

Model Coefficients (Unscaled):
MedInc      0.444758
AveRooms   -0.028150
HouseAge    0.016836
dtype: float64

Model Intercept (Unscaled):
0    0.017296
dtype: float64


In [68]:
coef_series_2 = pd.Series(lin_reg_selected.coef_, index=X_selected.columns)
intercept_2 = lin_reg_selected.intercept_

print("Coefficients (Unscaled):")
print(coef_series_2)
print("\nIntercept:", intercept_2)

Coefficients (Unscaled):
MedInc      0.444758
AveRooms   -0.028150
HouseAge    0.016836
dtype: float64

Intercept: 0.017295645283370575


In [69]:
# This row's feature values will be multiplied by our coefficients.
row_index_Selected = 1  # for demonstration
row_features_2 = X_selected.iloc[row_index_Selected]  # features only
print("Feature values (Row", row_index_Selected, "):\n", row_features_2)

Feature values (Row 1 ):
 MedInc       8.301400
AveRooms     6.238137
HouseAge    21.000000
Name: 1, dtype: float64


In [71]:
manual_prediction_selected = (row_features_2 * coef_series_2).sum() + intercept_2
print("\nManual Prediction for Row", row_index_Selected, ":", manual_prediction_selected)


Manual Prediction for Row 1 : 3.887367352235725


In [64]:
model_prediction_selected = lin_reg_selected.predict([row_features_2])
print("Model Prediction from lin_reg_selected.predict():", model_prediction_selected[0])

Model Prediction from lin_reg_selected.predict(): 3.887367352235725


/opt/miniconda3/lib/python3.13/site-packages/sklearn/utils/validation.py:2691: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [ ]:
import pandas as pd

comparison = pd.DataFrame({
    "Actual": y_test.values,
    "Predicted": y_pred_selected
})

print(comparison.head(10))   # shows first 10 comparisons


    Actual  Predicted
0  0.47700   1.067919
1  0.45800   1.506341
2  5.00001   2.328626
3  2.18600   2.681850
4  2.78000   2.091824
5  1.58700   2.168402
6  1.98200   2.709365
7  1.57500   2.158883
8  3.40000   2.110601
9  4.46600   4.189692


### Interpretation Questions
The simplified model clearly performs worse than the full model. Its R² drops from 0.58 to 0.49, meaning it explains less of the variation in housing prices, and its RMSE rises from 0.75 to 0.81, showing that its predictions are less accurate. The comparison between actual and predicted values also shows larger errors in the simplified version. Even though it is easier to interpret, I wouldn’t use this simplified model in practice because the loss in accuracy is significant, and the full model provides more reliable predictions.
